### PR TITLE
Matching algorithm

### PR DESCRIPTION
**Pull request by Xelowak**
Add a matching algorithm that works with users' accounts in the database.

For now, here are the criteria for an other user (O) to be presented to the current user (C):
1. O's gender corresponds to C's show me
2. At least one passion of C is in common with O
3. O is in the location range of C
4. O is in the age range of C, as specified in settings

Everything is then reverse checked (O and C are swapped)

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/blindly/match/Profile.kt [hunk 1]
```diff
@@ -2,6 +2,7 @@ package ch.epfl.sdp.blindly.match
 
 import android.os.Parcel
 import android.os.Parcelable
+import ch.epfl.sdp.blindly.user.User
```
Comment (cgntnr) [inline app/src/main/java/ch/epfl/sdp/blindly/match/Profile.kt:5]: Why was this import needed without any change in the file?

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                    filteredList += user
+                }
+            }
+        }
+
+        return filteredList
+    }
+
+    /**
+     * Does the same checks than [MatchingAlgorithm.getPotentialMatchesFromDatabase], but from the
```
Comment (cgntnr) [inline app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:43]: If they are doing the same checks wouldn't it be possible to merge these two and simply reverse the order of arguments passed. Though not sure if that would break anything, correct functioning comes first. I leave it up to you.

File: app/src/main/java/ch/epfl/sdp/blindly/user/User.kt [hunk 3]
```diff
@@ -81,7 +83,7 @@ class User private constructor(
          * @param location the location of the User
          */
         fun setLocation(location: List<Double>) = apply {
-            if(location.size == SIZE_2)
+            if (location.size == SIZE_2)
```
Comment (cgntnr) [inline app/src/main/java/ch/epfl/sdp/blindly/user/User.kt:86]: I wrote that in a different PR but I still believe SIZE_2 defeats the purpose of making it a constant.

**Review by cgntnr — APPROVED** [review]
Great job! Added couple of comments, nothing big.

File: app/src/main/java/ch/epfl/sdp/blindly/match/Profile.kt (hunk 1 above)
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/match/Profile.kt:5]: It was when I tried to implement a function in this class. I deleted it, so I need to delete this import too. Thanks for seeing it !

Comment (cgntnr) [reply app/src/main/java/ch/epfl/sdp/blindly/match/Profile.kt:5]: No worries !

File: app/src/main/java/ch/epfl/sdp/blindly/user/User.kt (hunk 3 above)
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/user/User.kt]: True ! I changed it.

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt (hunk 2 above)
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:43]: Unfortunately, it wouldn't be possible since `getPotentialMatchesFromDatabase` queries a Firebase Firestore database (see in `MatchingAlgorithm.kt`), which is not possible in this class (because it uses a list of users instead).

Comment (cgntnr) [reply app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:43]: I see, just wanted to check. Thanks for looking into it.

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @param userList the list to be filtered
+     * @return the same list without null users
+     */
+    fun clearNullUsers(userList: List<User?>): List<User> {
+        val nonNullList: MutableList<User> = ArrayList<User>().toMutableList()
+        for (user in userList) {
+            if (user != null) {
+                nonNullList += user
+            }
```
Comment (Lollipoke) [inline app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:83]: I think you can use filterNotNull for that ^^
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/filter-not-null.html

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        for (user in otherUsers) {
+            val otherUserAge = User.getUserAge(user)
+            if (otherUserAge != null) {
+                if (otherUserAge in minAge!!..maxAge!! && isLocatedInUserRadius(
+                        currentUser,
+                        user
+                    )
+                ) {
+                    filteredList += user
+                }
```
Comment (Lollipoke) [inline app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:35]: I think you can use the filter method for that, to make it cleaner
For example : https://stackoverflow.com/questions/44518251/list-filter-within-the-outer-range-kotlin

File: app/src/main/java/ch/epfl/sdp/blindly/match/MatchingAlgorithm.kt [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            query.get().addOnSuccessListener { users ->
+                for (user in users) {
+                    matches += user.toUser()
+                }
+            }.addOnFailureListener { exception ->
+                Log.w(TAG, "Error getting users : ", exception)
+            }
+        }
+
+        val nonNullMatches = userListFilter.clearNullUsers(matches)
```
Comment (Lollipoke) [inline app/src/main/java/ch/epfl/sdp/blindly/match/MatchingAlgorithm.kt:69]: Here you could use the filterNotNull method in one line, as I said in the other comment

File: app/src/androidTest/java/ch/epfl/sdp/blindly/match/UserListFilterTest.kt [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    @Test
+    fun filterLocationAndAgeRangeWorks() {
+        val alice = userBuilder.setUsername("Alice")
+            .setBirthday("29.04.1997")
+            .setAgeRange(listOf(20, 30))
+            .setLocation(locationEPFL)
+            .setRadius(50)
+            .build()
+
+        // No problem, Bob is perfect for Alice ! <3
```
Comment (Lollipoke) [inline app/src/androidTest/java/ch/epfl/sdp/blindly/match/UserListFilterTest.kt:36]: Yaaaay, gotta love the incoming love story <3

**Review by Lollipoke — COMMENTED** [review]
I think it would be cleaner and easier to use the built-in filter methods for some function you wrote, especially if we aim for an efficient algorithm. Other than that, it looks clean, very well documented and well writen, great job !

File: app/src/main/java/ch/epfl/sdp/blindly/user/User.kt [hunk 8]
```diff
@@ -81,7 +81,7 @@ class User private constructor(
          * @param location the location of the User
          */
         fun setLocation(location: List<Double>) = apply {
-            if(location.size == SIZE_2)
+            if (location.size == 2)
```
[the remark this answers is not in the document]
Comment (cgntnr) [reply app/src/main/java/ch/epfl/sdp/blindly/user/User.kt:84]: It was due to our previous exchange, he changed that to 2, because it was SIZE_2 which is not very meaningful either. I think we had a communication breakdown there what I meant was maybe to put something more explanatory like SIZE_OF_LOC_ATTRIBUTES etc. Because ```location.size == SIZE_2``` is not really better than ```location.size == 2```.

[the remark this answers is not in the document]
Comment (Lollipoke) [reply app/src/main/java/ch/epfl/sdp/blindly/user/User.kt:84]: yeah maybe the name could be more meaningful indeed

File: app/src/main/java/ch/epfl/sdp/blindly/main_screen/MatchPageFragment.kt [hunk 9]
```diff
@@ -173,6 +173,41 @@ class MatchPageFragment : Fragment(), CardStackListener {
...
+        } else {
+            createProfilesFromUsers(potentialUsers)
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun createProfilesFromUsers(users: List<User>?): List<Profile> {
+        if (users == null) {
+            return listOf()
+        }
```
[the remark this answers is not in the document]
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/main_screen/MatchPageFragment.kt:200]: Yes it is, since the function takes a `List<User>?`, which can be null. Otherwise, I could use the "?" notation, but I think this null test handles this better.

File: app/src/main/java/ch/epfl/sdp/blindly/user/User.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/user/User.kt:84]: At the end, I used two constants with differents names, namely `SIZE_OF_LOCATION_LIST` and `SIZE_OF_AGE_RANGE_LIST` which are both equal to 2.

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt (hunk 4 above)
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:83]: True, I didn't know something existed for that.

File: app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt (hunk 5 above)
Comment (Xelowak) [reply app/src/main/java/ch/epfl/sdp/blindly/match/UserListFilter.kt:35]: Yep, I'll try it.

**Review by Lollipoke — APPROVED** [review]
